File: LeetCode-DSA/JulyChallenge_2022/MiddleLinkedList.cpp
/* Q876. Description: Given the head of a singly linked list, return the middle node of the linked list.
If there are two middle nodes, return the second middle node.
 */

class Solution {
public:
    ListNode* middleNode(ListNode* head) {
        ListNode* fast = head;
        ListNode* slow = head;
        while(fast != NULL && fast -> next != NULL)
        {
            slow = slow -> next;
            fast = fast -> next -> next;
        }
        return slow;
    }
};

File: LeetCode-DSA/LeetCode75/Level1/MiddleLinkedList.java
/* 
Q876. Description: Given the head of a singly linked list, return the middle node of the linked list.

If there are two middle nodes, return the second middle node.
 */

class Solution {
    public ListNode middleNode(ListNode head) 
    {
        ListNode fast = head;
        ListNode slow = head;
        
        // When fast reaches end of list, slow is the middle
        while(fast != null && fast.next != null)
        {
            // Fast travels 2 elements at a time
            fast = fast.next.next;
            slow = slow.next;
        }
        return slow;
    }
}